### PR TITLE
Adding tables for top 10 container producing/collecting logs [Log-1270] [WIP]

### PR DESCRIPTION
**Pull request by ajaygupta978**
### Description
This PR  adds  tables that displays the top 10 containers of logs produced and logs collected. 

/cc  @alanconway 
/assign @jcantrill 

### Links

- Depending on PR(s): https://github.com/openshift/cluster-logging-operator/pull/979, https://github.com/openshift/origin-aggregated-logging/pull/2070
- Bugzilla:
- Github issue: 
- JIRA: [Log-1270](https://issues.redhat.com/browse/Log-1270)
- Enhancement proposal:

### CODE REVIEW

**Comment by ajaygupta978** [comment]
Here is screen shot:
Screenshot 2021-04-29 at 1 05 29 AM


@alanconway /@jcantrill :  Could you please review it ?
I have two  questions:
1) In the JIRA ticket it was mentioned to add only one table which shows logs produced. Since we have two charts, one for logs produced and one for logs collected so I added another table to show logs collected. Please let me know if its ok.
2) In the table I am able to add pod/container name but somehow query is not working to add `namespace`.  Though pod name shows namespace also  so all information is in the table but I am still trying to figure out if namespace  can be added as separate column.

This PR is dependent on @pmoogi-redhat PR so lets wait before merging it. 

cc: @sichvoge

**Comment by jcantrill** [comment]
> I have two questions:
> 
> 1. In the JIRA ticket it was mentioned to add only one table which shows logs produced. Since we have two charts, one for logs produced and one for logs collected so I added another table to show logs collected. Please let me know if its ok.

@sichvoge did we not explicitly state we did not want a table for "collected" so that it is less obvious that we have missed logs?  Or was it we did not want a table/chart to identify the gap?

> 2. In the table I am able to add pod/container name but somehow query is not working to add `namespace`.  Though pod name shows namespace also  so all information is in the table but I am still trying to figure out if namespace  can be added as separate column.

what is your question?

File: files/dashboards/openshift-logging-dashboard.json
Comment (alanconway) [inline files/dashboards/openshift-logging-dashboard.json:1840]: 5m seems short. 24 hours might be a more useful number, since the logging activity is likely to vary between night and day. Retention times are in days, weeks or months so a day seems like a reasonable unit to project from.

**Review by alanconway — CHANGES_REQUESTED** [review]
Can you attach some screenshots also? I can't easily read the dashboard code.

**Comment by ajaygupta978** [comment]
Screenshot 2021-04-29 at 1 54 00 AM
Screenshot 2021-04-29 at 1 53 54 AM

@alanconway Please see attached screenshots

**Comment by ajaygupta978** [comment]
> > I have two questions:
> > 
> > 1. In the JIRA ticket it was mentioned to add only one table which shows logs produced. Since we have two charts, one for logs produced and one for logs collected so I added another table to show logs collected. Please let me know if its ok.
> 
> @sichvoge did we not explicitly state we did not want a table for "collected" so that it is less obvious that we have missed logs? Or was it we did not want a table/chart to identify the gap?
> 
> > 1. In the table I am able to add pod/container name but somehow query is not working to add `namespace`.  Though pod name shows namespace also  so all information is in the table but I am still trying to figure out if namespace  can be added as separate column.
> 
> what is your question?
Is it must to have  `namespace` column in table? Query requires group by two columns but not sure if OCP support it.

File: files/dashboards/openshift-logging-dashboard.json
Comment (ajaygupta978) [reply files/dashboards/openshift-logging-dashboard.json:1840]: @alanconway I think we can have dynamic value based on drop down. From 5 mins to 2 hours charts always shows values based on last 5 mins but beyond that it uses actual selected value which varies from 6 hours, 12 hours , 1 day,1 week, 2 weeks.

Comment (ajaygupta978) [reply files/dashboards/openshift-logging-dashboard.json]: I have added time-interval support so now user can select time-interval from drop down. Based on selected time-interval table values will be reflected.

**Comment by alanconway** [comment]
> > > I have two questions:
> > > 
> > > 1. In the JIRA ticket it was mentioned to add only one table which shows logs produced. Since we have two charts, one for logs produced and one for logs collected so I added another table to show logs collected. Please let me know if its ok.
> > 
> > 
> > @sichvoge did we not explicitly state we did not want a table for "collected" so that it is less obvious that we have missed logs? Or was it we did not want a table/chart to identify the gap?

My memory is we eliminated the chart showing the gap, but provide all the information so customers can do the calculation if they choose.

> > > 1. In the table I am able to add pod/container name but somehow query is not working to add `namespace`.  Though pod name shows namespace also  so all information is in the table but I am still trying to figure out if namespace  can be added as separate column.

Both the Go and fluentd colllectors are supposed to provide the labels: namespace, pod, uid (of pod), container. Those should be the headings instead of "path".

**Comment by jcantrill** [comment]
> 
> Both the Go and fluentd colllectors are supposed to provide the labels: namespace, pod, uid (of pod), container. Those should be the headings instead of "path".

This info can also be parsed from the file path if not already as a label.  @ajaygupta978 please make an attempt to create a table with the following columns in this order:

* namespace
* pod
* container hash
* total bytes collected

sorted by total bytes collected

**Comment by ajaygupta978** [comment]
@jcantrill This information is available and I am able to create top 10 table with Pod, PodName, TotalBytes with `topk(10,sum(increase(log_logged_bytes_total[$__interval])) by (path, pod, podname))"` query.  But if add `namespace` column in above query like  `topk(10,sum(increase(log_logged_bytes_total[$__interval])) by (path, pod, podname, namespace))"`  I am getting only 1 row. This query works fine in prometheus dashboard but not returning expected results on dashboard.

**Comment by ajaygupta978** [comment]
@jcantrill: As I mentioned there was bug (https://bugzilla.redhat.com/show_bug.cgi?id=1970476) in OCP so based on your suggestion I found  a workaround. I parsed `path` and extracted `namespace` and renamed to `znamespace` and got following table (screen-shot) 
FinalTable
It doesn't have container hash. I am already using label_replace for extracting namespace so not sure if its possible to extract hash. Also sorting is not supported according to ocp team but user can click and sort based on `Total Value` column. 

There is one more issue in OCP as of now. In dashboard json we use `$__interval` variable which represent time interval to calculate metric. This is populated based on user selected value from drop down. But table value doesn't change if user selects any of the value between [5m-6h] and beyond that it does but values are  not accurate. Based on discussion with OCP this seems to be another bug. 
We can decide a workaround to hardcode interval value to `1h` as of now which gives total bytes logged/collected in past one hour. 
1. Please suggest if this temporary workaround is good to go?
2. Also if I push changes now I am assuming it will not be merged because it depends on log-loss metrics so QE team can only test post merging log-loss PR so should I push now or wait for other PR to merge and test again ?

**Comment by ajaygupta978** [comment]
@jcantrill As we discussed somehow this PR commits got corrupted so I created new [PR-1056](https://github.com/openshift/cluster-logging-operator/pull/1056) for merging. We can close this once we merge new PR.

**Comment by jcantrill** [comment]
> @jcantrill: As I mentioned there was bug (https://bugzilla.redhat.com/show_bug.cgi?id=1970476) in OCP so based on your suggestion I found a workaround. I parsed `path` and extracted `namespace` and renamed to `znamespace` and got following table 
> 
> There is one issue in OCP as of now. In dashboard json we use `$__interval` variable which represent time interval to calculate metric. This is populated based on user selected value from drop down. But table value doesn't change if user selects any of the value between [5m-6h] and beyond that it does but values are not accurate. Based on discussion with OCP this seems to be another bug. They asked to create a bug.
> I observed other charts also using pre-defined values. We can decide a workaround to hardcode interval value to `1h` as of now which gives total bytes logged/collected in past one hour.
> 
> 1. Please suggest if this temporary workaround is good to go?

I defer to the default @alanconway recommended previously.  Is it possible for us to use "interval" for now and it will be "fixed" when the bz is fixed?  Or is the issue that we do not get any sane default

> 2. Also if I push changes now I am assuming it will not be merged because it depends on log-loss metrics so QE team can only test post merging log-loss PR so should I push now or wait for other PR to merge and test again ?

This is correct.  Additionally, please clean up the PR as I would expect it to only have a single commit with only the changes related to your work.

**Comment by jcantrill** [comment]
closing in favor of #1056

**Comment by ajaygupta978** [comment]
For `$__interval` issue [BZ bug](https://bugzilla.redhat.com/show_bug.cgi?id=1973406) is raised.

**Pull request closed without merging**
